clinica/pipelines/dwi_connectome/dwi_connectome_pipeline.py

```diff
@@ -56,75 +56,98 @@ def build_input_node(self):
-        error_message += check_input_caps_files(aparc_aseg_files, "T1_FS_DESIKAN", "t1-freesurfer",
-                                                self.caps_directory, self.subjects, self.sessions)
+        aparc_aseg_files, err_msg = clinica_file_reader(self.subjects,
+                                                        self.sessions,
+                                                        self.caps_directory,
+                                                        {'pattern': 'sub-*_ses-*/mri/aparc+aseg.mgz',
+                                                         'description': 'Desikan-based segmentation (mri/aparc.a2009s+aseg.mgz)',
+                                                         'needed_pipeline': 't1-freesurfer'})
+        if err_msg:
+            all_errors.append(err_msg)
 
         # Destrieux parcellation
-        aparc_aseg_a2009s_files = caps_layout.get(freesurfer_file='aparc.a2009s\+aseg.mgz', return_type='file',
-                                                  subject=participants_regex, session=sessions_regex)
-        error_message += check_input_caps_files(aparc_aseg_a2009s_files, "T1_FS_DESTRIEUX", "t1-freesurfer",
-                                                self.caps_directory, self.subjects, self.sessions)
+        aparc_aseg_a2009s_files, err_msg = clinica_file_reader(self.subjects,
+                                                               self.sessions,
+                                                               self.caps_directory,
+                                                               {'pattern': 'sub-*_ses-*/mri/aparc.a2009s+aseg.mgz',
+                                                                'description': 'Destrieux-based segmentation (mri/aparc.a2009s+aseg.mgz)',
+                                                                'needed_pipeline': 't1-freesurfer'})
+        if err_msg:
+            all_errors.append(err_msg)
 
         # Inputs from dwi-preprocessing pipeline
         # ======================================
         # Preprocessed DWI
-        dwi_files = caps_layout.get(type='dwi', suffix='preproc', extensions=['.nii|.nii.gz'], return_type='file',
-                                    subject=participants_regex, session=sessions_regex)
-        error_message += check_input_caps_files(dwi_files, "DWI_PREPROC_NII", "dwi-preprocessing",
-                                                self.caps_directory, self.subjects, self.sessions)
+        dwi_files, err_msg = clinica_file_reader(self.subjects,
+                                                 self.sessions,
+                                                 self.caps_directory,
+                                                 {'pattern': '*_dwi_space-*_preproc.nii*',
+                                                  'description': 'preprocessed DWI',
+                                                  'needed_pipeline': 'dwi-preprocessing-using-t1 or dwi-preprocessing-using-fieldmap'})
```

Comment:
What about using a dictionary instead?

When you will modify `dwi-dti`, you will copy/paste these informations and if one day, the CAPS, the description or the name of the pipeline is changing, you will have to change each line where the file is grabbed instead of modifying the dictionary once.

Reply:
I'd rather duplicate a bit of code for the following reasons:
- The number of occurrences of grabbing the same file is quite low
- I find it very convenient to have all the required information to make the clinica_file_reader call without external call 
- For the same reason, it is then easier to maintain 
- CAPS changing names or pipeline changing name is (kind of) unlikely
- If this ever happens, it is really a matter of seconds to change it with Cmd + Shift + F in Pycharm (see below)
![Screen Shot 2019-10-11 at 14 02 44](https://user-images.githubusercontent.com/49677712/66650004-d86ab280-ec2f-11e9-990e-bba2cffdf6e0.png)

Reply:
> * The number of occurrences of grabbing the same file is quite low

This is not a good argument - DRY (Don’t Repeat Yourself):
> If you have identical code in 3 or more places, it should be abstracted into a common method, class, etc. to be reusable and easily maintainable.

 :wink:


> * I find it very convenient to have all the required information to make the clinica_file_reader call without external call
> * For the same reason, it is then easier to maintain

You can have this behaviour using e.g. dictionaries, enum or classes mechanism.


> * CAPS changing names or pipeline changing name is (kind of) unlikely
> * If this ever happens, it is really a matter of seconds to change it with Cmd + Shift + F in Pycharm

It will eventually change when CAPS will be merged into BIDS-Derivatives. I agree that it is fast but it easier to modify the regexp once for all than to modify N times the regexp (e.g. 6 for the example with T1w files)

Before that, it is more likely that a new pipeline will be created and you will need to grab files. In this scenario, it will be very complicated to enforce that the developer uses the same description if the same files have been grabbed in another pipeline.

Reply:
If my comment regarding grabbing files from `t1-freesurfer` pipeline is correct, you can see than you need to modify each pipeline to fix the regexp (i.e. `fmri-pipeline`, `pet-surface`, `dwi-connectome`).

In general, each possible error in a regexp  currently needs to be fixed into the involved pipeline(s).